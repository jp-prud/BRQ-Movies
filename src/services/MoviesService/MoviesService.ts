import { FavoriteMovieProps, MovieProps, MoviePropsAPI, Page, PageAPI, StorageKeys } from "@types"
import { storage } from "../StorageService/storage"
import { HttpClient } from "../utils/HttpClient"
import { APIMapper } from "../utils/mappers/APIMapper"
import { MovieMapper } from "../utils/mappers/MovieMapper"

export function MoviesService() {
  async function listMovies(page: number = 1): Promise<Page<MovieProps>> {
    const { toPageModel }  = APIMapper()
    const { data } = await HttpClient.get<PageAPI<MoviePropsAPI>>(`/movie/popular?language=en-US&page=${page}`)
    return toPageModel(data, MovieMapper().toPresentation)
  }

  function listFavoriteMovies(): Promise<FavoriteMovieProps[] | null> { 
    return storage.getItem<FavoriteMovieProps[]>(StorageKeys.FavoriteMovies)
  }

  async function getMovieById(movieId: MovieProps['id']): Promise<MovieProps> { 
    const { toPresentation } = MovieMapper()
    const { data } = await HttpClient.get<MoviePropsAPI>(`/movie/${movieId}?language=en-US`)
    return toPresentation(data)
  }

  async function favoritedMovieById(movieId: MovieProps['id']): Promise<FavoriteMovieProps | null> { 
    const favoriteMovies = await storage.getItem<FavoriteMovieProps[]>(StorageKeys.FavoriteMovies) || [];
    return favoriteMovies.find(movie => movie.id === movieId) || null
  }

  async function favoriteMovie(favoriteMovie: Pick<MovieProps, 'id' | 'posterPath'>): Promise<void> { 
    const favoriteMovies = await storage.getItem<FavoriteMovieProps[]>(StorageKeys.FavoriteMovies) || [];

    const isFavorited = favoriteMovies.some(movie => movie.id === favoriteMovie.id)

    if (isFavorited) { 
      const newFavoriteMovies = favoriteMovies.filter(movie => movie.id !== favoriteMovie.id)
      storage.setItem(StorageKeys.FavoriteMovies, newFavoriteMovies)

      return
    }

    const newFavoriteMovies = [...favoriteMovies, favoriteMovie]
    storage.setItem(StorageKeys.FavoriteMovies, newFavoriteMovies)
  }

  async function getSimilars(movieId: MovieProps['id'], page: number = 1): Promise<Page<MovieProps>> {
    console.log('movieId', movieId)

    const { toPageModel }  = APIMapper()
    const { data } = await HttpClient.get<PageAPI<MoviePropsAPI>>(`/movie/${movieId}/similar?language=en-US&page=${page}`)
    return toPageModel({
      ...data,
      results: data.results.slice(0, 20)
    }, MovieMapper().toPresentation)
  }

  return {
    listMovies,
    listFavoriteMovies,
    favoriteMovie,
    favoritedMovieById,
    getMovieById,
    getSimilars
  }
}