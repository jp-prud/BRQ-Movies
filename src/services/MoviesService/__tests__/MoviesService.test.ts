import { server } from '@tests';

import { MoviesService } from '../MoviesService';

beforeAll(() => {
  server.listen();
});

afterEach(() => {
  server.resetHandlers();
  jest.clearAllMocks();
});

afterAll(() => {
  server.close();
});

describe('MoviesService', () => {
  it('should return a list of movies', async () => {
    const { listMovies } = MoviesService();

    const movies = await listMovies();

    expect(movies.data.length).toBeGreaterThan(0);
    expect(movies.data.length).toBe(3); // Ajuste de acordo com os dados mockados esperados
    // Verificar a estrutura dos filmes
    movies.data.forEach(movie => {
      expect(movie).toHaveProperty('id');
      expect(movie).toHaveProperty('title');
    });
  });

  it("should render movie's details", async () => {
    const { listMovies, getMovieById } = MoviesService();

    const movies = await listMovies();

    const movie = await getMovieById(movies.data[0].id);

    expect(movie).toHaveProperty('title', movies.data[0].title);
    expect(movie).toHaveProperty('id', movies.data[0].id);
  });

  it('should render favorite movies', async () => {
    const { listMovies, favoriteMovie, listFavoriteMovies } = MoviesService();

    const movies = await listMovies();
    const movie = movies.data[0];

    await favoriteMovie(movie);

    const favoriteMovies = await listFavoriteMovies();

    expect(favoriteMovies).toEqual([movie]);
    expect(favoriteMovies.length).toBe(1); // Removido o operador "!"
  });

  it('should favorite a movie', async () => {
    const { listMovies, favoriteMovie, listFavoriteMovies } = MoviesService();

    const movies = await listMovies();
    const movie = movies.data[0];

    await favoriteMovie(movie);

    const favoriteMovies = await listFavoriteMovies();

    expect(favoriteMovies).toMatchSnapshot();
  });

  test('is favorited movie', async () => {
    const { listMovies, favoriteMovie, favoritedMovieById } = MoviesService();

    const movies = await listMovies();
    const movie = movies.data[0];

    await favoriteMovie(movie);

    const favoritedMovie = await favoritedMovieById(movie.id);

    expect(movie).toEqual(favoritedMovie);
  });
});
