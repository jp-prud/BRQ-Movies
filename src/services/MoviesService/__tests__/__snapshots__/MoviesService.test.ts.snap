// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesService should favorite a movie 1`] = `[]`;
